src/services/redisSessionStorage.ts

```diff
@@ -2,64 +2,187 @@
-    const client = this.redisClient;
+  public get(token: string): Promise<Either<Error, ISessionState>> {
+    return new Promise(resolve => {
+      // Get the fields for this key.
+      // @see https://redis.io/commands/hmget
+      this.redisClient.hgetall(token, (err, value) => {
+        if (err) {
+          // Client returns an error.
+          return resolve(left<Error, ISessionState>(new Error(err.message)));
+        }
+
+        if (value === null || value === undefined) {
+          // Null and undefined values are considered "not found".
+          return resolve(this.sessionNotFound());
+        }
 
+        // Check if the token has expired. We don't remove expired token
+        // because a client can later refresh the session.
+        const timestamp = +value.timestamp;
+        if (timestamp + this.tokenDurationSecs * 1000 < Date.now()) {
+          return resolve(
+            left<Error, ISessionState>(new Error("Token has expired"))
+          );
+        }
+
+        // We got a value, let's see if it's a valid User.
+        const errorOrUser = extractUserFromJson(value.data);
+
+        errorOrUser.fold(
+          // Not a valid User, return an error.
+          () => resolve(this.unableToDecodeUser()),
+
+          // We got a valid user from the session.
+          user =>
+            resolve(
+              right<Error, ISessionState>({
+                expireAt: timestamp + this.tokenDurationSecs * 1000,
+                expired: false,
+                user
+              })
+            )
+        );
+      });
+    });
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public refresh(token: string): Promise<Either<Error, ISessionState>> {
     return new Promise(resolve => {
-      // Get the value of key.
-      // @see https://redis.io/commands/get
-      client.get(token, (err, value) => {
+      // Get the fields for this key.
+      // @see https://redis.io/commands/hgetall
+      this.redisClient.hgetall(token, (err, value) => {
         if (err) {
-          resolve(left<Error, User>(new Error(err.message)));
-        } else {
-          if (value === null || value === undefined) {
-            resolve(left<Error, User>(new Error("Session not found")));
-          } else {
-            const errorOrUser = extractUserFromJson(value);
-
-            // TODO: better error message.
-            errorOrUser.mapLeft(() => {
-              return "Errors in validating the user profile";
-            });
-
-            resolve(errorOrUser);
-          }
+          // Client returns an error.
+          return resolve(left<Error, ISessionState>(err));
+        }
+
+        if (value === null || value === undefined) {
+          // Null and undefined values are considered "not found".
+          return resolve(this.sessionNotFound());
         }
+
+        // We got a value, let's see if it's a valid User.
+        const errorOrUser = extractUserFromJson(value.data);
```

Comment:
since this is basically the same logic implemented in `get`, it could make sense to wrap `redisClient.hgetall` into a function that returns `Promise<Either<Error, User>>`